### PR TITLE
Add solution 9 - a demo of working with Timers in Orleans.

### LINKED ISSUE #149: Illustrate timers.

### PR DESCRIPTION
**Pull request by CamiloTerevinto**
Aims to fix #149

### CODE REVIEW

**Comment by CamiloTerevinto** [comment]
Main files are: 

* Controller: https://github.com/PiotrJustyna/road-to-orleans/blob/806fd1fe11d65cc9bf5ec84a0f86f108f16c273c/9/Api/src/Controllers/TimerWorldController.cs 
* Grain: https://github.com/PiotrJustyna/road-to-orleans/blob/806fd1fe11d65cc9bf5ec84a0f86f108f16c273c/9/Grains/src/TimerWorld.cs
* Grain unit tests: https://github.com/PiotrJustyna/road-to-orleans/blob/806fd1fe11d65cc9bf5ec84a0f86f108f16c273c/9/Grains.UnitTests/TimerUnitTest.cs

**Comment by PiotrJustyna** [comment]
Thanks for taking the initiative Camilo. I should have some time next week to review this.

File: 9/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+**/charts
+**/docker-compose*
+**/Dockerfile*
+**/node_modules
+**/npm-debug.log
+**/obj
+**/secrets.dev.yaml
+**/values.dev.yaml
+LICENSE
+README.md
```
Comment (PiotrJustyna) [inline 9/.dockerignore:25]: Not sure we'll need this gitignore. We already have one on the root level: https://github.com/PiotrJustyna/road-to-orleans/pull/202/files

Comment (PiotrJustyna) [reply 9/.dockerignore:25]: Looks like solution 6 also has as gitignore (https://github.com/PiotrJustyna/road-to-orleans/blob/main/6/.dockerignore), creating an issue to remove them from individual solutions.

Comment (PiotrJustyna) [reply 9/.dockerignore:25]: Done: https://github.com/PiotrJustyna/road-to-orleans/issues/203

File: 9/Grains/src/TimerWorld.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+            _logger.LogInformation($"{nowInSeconds}: This timer has executed {_executionCount} times - {IdentityString}");
+
+            // this is just to demonstrate that the timer waits for `period`
+            // after the execution finishes before starting the next one
+            // meaning that the log above will happen every 5 seconds
+            await Task.Delay(4000);
+        }
+    }
+}
```
Comment (PiotrJustyna) [inline 9/Grains/src/TimerWorld.cs:101]: Oh, I see what you're doing. This is great and I like the fact that we're registering a timer on grain method execution and not `OnActivateAsync`, but I'm wondering, is it not too much for a simple `Timers` example? The idea I have for this repo is to trim all the fat we don't need to get an example's point across and here we have:

* Web API
* old Web API examples (the whole `hello world` path)
* grain deactivation
* etc.

What I had in mind was something super simple and limited in scope, like an example I wrote sometime earlier last year: https://github.com/PiotrJustyna/orleans-sandbox/tree/main/hello-world-timer

So my thinking is:

* timers are quite easy to set up, so they shouldn't need too much code/configuration around them
* orleans timers aren't actually so trivial to work with and they are not exactly working the same way as the regular timers do:
  * asynchronous callbacks cannot overlap so their executions can delay each other (https://dotnet.github.io/orleans/docs/grains/timers_and_reminders.html#timer-usage). I illustrated it in my example (linked in this comment, not sure if I succeeded getting the point across, though, so feel free to improve it)
  * grains are single-threaded which can lead to interesting observations
  * timer registration actually should not be done at construction time, but rather on activation or grain method execution, which your example partially demonstrates

So in general, I think what would be most helpful for our readers is:

* dramatically less code surrounding the actual demo (your new grain) - what I would suggest is to take e.g. Solution 3 as your starting point and create e.g. Solution 3b.
* really interesting examples illustrating the gotchas of orleans timers as compared to the regular .net timers
* maybe we could use a grain per timer example to keep the examples well isolated - don't want the grain code to get too messy

**Review by PiotrJustyna — COMMENTED** [review]
Thanks again for all the hard work you've invested into this. I left some comments, please let me know what you think.

Also, if you ever want to update the diagram of all solutions, I used Plant UML for it - you can get a VS Code plugin for rendering and exporting diagrams. Current version of the diagram is here: https://github.com/PiotrJustyna/road-to-orleans/blob/main/solutions.puml

File: 9/run-demo-load-test.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+docker build -t vegeta-loadtest -f ./ops/Vegeta/Dockerfile ./ &&
+  docker run -it --rm vegeta-loadtest
```
Comment (PiotrJustyna) [inline 9/run-demo-load-test.sh:3]: These load test scripts should probably be revised. The vegeta script in this solution is targeting one of the older controllers (hello world). Probably best to remove the load test altogether.

Also, the `/imgs` files are for a different solution and should ideally be removed too.

File: readme.md [hunk 4]
```diff
@@ -32,6 +32,7 @@ Silo dashboards are available by default on `localhost:8080` unless configured o
 | [solution6](6/readme.md) | Solution5 where the cluster is being called from a Web API. | :white_check_mark: | in-memory | n | 1 C# grain interfacing F# library code | n - web api |||
 | [solution7](7/readme.md) | Solution6 + [FeatureManagement](https://www.nuget.org/packages/Microsoft.FeatureManagement/), dependency injection in grains, unit tests for grains using [OrleansTestKit](https://www.nuget.org/packages/OrleansTestKit/). | :white_check_mark: | in-memory | n | 1 C# grain interfacing F# library code | n - web api |||
 | [solution8](8/readme.md) | Solution7 + Persistent Membership Table in DynamoEB, ECS hosting, CloudFormation Template. | :white_check_mark: | Dynamo | n | 1 C# grain interfacing F# library code | n - web api |||
+| [solution9](9/readme.md) | Solution7 + Timers. | :white_check_mark: | in-memory | n | 2 C# grains interfacing F# library code | n - web api |||
```
Comment (PiotrJustyna) [inline readme.md:35]: Is it not based on solution 6?

File: 9/readme.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+* work with [Timers](https://dotnet.github.io/orleans/docs/grains/timers_and_reminders.html)
+
+## running the code
+
+### silos
+
+As in example 6, following scripts can be used to run silos:
+
+* `./run-docker-silo.sh`
+* `./run-local-silo.sh`
```
Comment (PiotrJustyna) [inline 9/readme.md:14]: Shouldn't it be `./run-silo-docker.sh` and `./run-silo-local.sh` instead?

File: 9/readme.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# readme
+
+This example builds on top of solution 7 and demonstrates how to:
```
Comment (PiotrJustyna) [inline 9/readme.md:3]: Does it not build on top of solution 6?

File: solutions.puml [hunk 7]
```diff
@@ -52,6 +56,7 @@ hide empty description
     Solution4 --> Solution5
     Solution5 --> Solution6
     Solution6 --> Solution7
+    Solution6 --> Solution9
     Solution7 --> Solution8
     Solution8 --> [*]
   }
```
Comment (PiotrJustyna) [inline solutions.puml:62]: Would you mind also rendering this for the main readme? `https://github.com/CamiloTerevinto/road-to-orleans/tree/feature/149/solutions`

File: readme.md (hunk 4 above)
Comment (PiotrJustyna) [reply readme.md:35]: Also, are you sure there are two grains? I only noticed one (`Timer`)?

**Review by PiotrJustyna — COMMENTED** [review]
Thanks for the PR, the change looks more compact than the last iteration. Left some comments, nothing major.

**Comment by CamiloTerevinto** [comment]
Hey @PiotrJustyna, all comments solved.

**Pull request closed without merging**